components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java

```diff
@@ -73,65 +87,166 @@ public AbstractHTTPFunction() {
-                        if (contentType != null && contentType.getValue().contains(TYPE_TEXT_PLAIN)) {
-                            // For 'text/plain', put the response body into the JSON object as a single field.
-                            json = new JSONObject();
-                            json.put(RESPONSE, jsonString);
-                        } else {
-                            JSONParser parser = new JSONParser();
-                            json = (JSONObject) parser.parse(jsonString);
+                if (!isValidRequestDomain(request.getURI())) {
+                    LOG.error("Provided Url does not contain a allowed domain. Invalid Url: " + endpointURL);
+                    asyncReturn.accept(context, Collections.emptyMap(), Constants.OUTCOME_FAIL);
+                } else {
+                    Pair<String, JSONObject> result = executeRequest(request, endpointURL);
+                    if (result.getLeft().equals(Constants.OUTCOME_FAIL)) {
+                        if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                            DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder = new
+                                    DiagnosticLog.DiagnosticLogBuilder(Constants.LogConstants.ADAPTIVE_AUTH_SERVICE,
+                                    Constants.LogConstants.ActionIDs.RECEIVE_API_RESPONSE);
+                            diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.EXTERNAL_API, endpointURL)
+                                    .resultMessage("Error while calling endpoint, proceeding with retry attempts.")
+                                    .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                                    .resultStatus(DiagnosticLog.ResultStatus.FAILED);
+                            LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
+                        }
+                        LOG.error("Error while calling endpoint. Url: " + endpointURL);
+                        result = executeRequestWithRetries(request, endpointURL, maxRequestAttemptsForAPIEndpointTimeout);
+                    } else {
+                        if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                            DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder = new
+                                    DiagnosticLog.DiagnosticLogBuilder(Constants.LogConstants.ADAPTIVE_AUTH_SERVICE,
+                                    Constants.LogConstants.ActionIDs.RECEIVE_API_RESPONSE);
+                            diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.EXTERNAL_API, endpointURL)
+                                    .resultMessage("Successfully called the external endpoint.")
+                                    .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                                    .resultStatus(DiagnosticLog.ResultStatus.SUCCESS);
+                            LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
```

Comment:
Can you point the log that user will see upon executing above  debug log?

Reply:
`2024-04-10 12:00:24,661|2a05a061-ef74-4208-bf96-b1a11c686de2|Tenant:carbon.super|pool-35-thread-1|{"logId":"2af6bc49-2d9b-4433-ab48-0995047f964e","recordedAt":{"seconds":1712730624,"nanos":660031000},"requestId":"2a05a061-ef74-4208-bf96-b1a11c686de2","resultStatus":"SUCCESS","resultMessage":"Successfully called the external endpoint.","actionId":"receive-api-response","componentId":"adaptive-auth-service","input":{"external api":"https://token.free.beeceptor.com/token"},"logDetailLevel":"APPLICATION"}`